File: ContainsDuplicate_optimal.py
from typing import List
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        hash_map = {}
        for i,num in enumerate(nums):
            if num in hash_map:
                return True
            else:
                hash_map[num] = i
        return False

sol = Solution()
print(sol.containsDuplicate([1,3,4,2,6]))
print(sol.containsDuplicate([1,3,4,2,1]))
print(sol.containsDuplicate([1,3,1]))
        

File: ContainsDuplicate_optimal.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        unordered_map<int, int> hashMap;

        for (int i = 0; i < nums.size(); i++) {
            if (hashMap.find(nums[i]) != hashMap.end()) {
                return true;
            }

            hashMap[nums[i]] = i;
        }

        return false;
    }
};

int main() {
    Solution sol;

    vector<int> nums1 = {1, 3, 4, 2, 6};
    vector<int> nums2 = {1, 3, 4, 2, 1};
    vector<int> nums3 = {1, 3, 1};

    cout << boolalpha;
    cout << sol.containsDuplicate(nums1) << "\n";
    cout << sol.containsDuplicate(nums2) << "\n";
    cout << sol.containsDuplicate(nums3) << "\n";

    return 0;
}
